kart/parse_args.py

```diff
@@ -75,29 +61,89 @@ def get_arg_type(repo, arg, allow_options=True, allow_commits=True, allow_filter
                 f"This will be removed in Kart 0.12! Please comment on {issue_link} if you need to use this option.",
                 RemovalInKart012Warning,
             )
-            return ArgType.OPTION
-        else:
-            raise click.UsageError(f"No such option: {arg}")
+            options.append(arg)
+    return options, others
+
+
+def _append_kwargs_to_options(options, kwargs, allow_options):
+    if not kwargs:
+        return
+    assert allow_options
+
+    for option_name, option_val in kwargs.items():
+        option_name = option_name.replace("_", "-", 1)
+        t = type(option_val)
+        if t == int or t == str:
+            options.append(f"--{option_name}={option_val}")
+        elif t == tuple:
+            options.extend([f"--{option_name}={o}" for o in option_val]),
+        elif t == bool and option_val:
+            options.append(f"--{option_name}")
 
-    if allow_commits:
-        if arg == "[EMPTY]" or ".." in arg:
-            return ArgType.COMMIT
 
-        try:
-            repo.resolve_refish(arg)
-            return ArgType.COMMIT
-        except (KeyError, ValueError, pygit2.InvalidSpecError):
-            pass
+HEAD_PATTERN = re.compile(r"^HEAD\b", re.IGNORECASE)
 
-    if allow_filters:
-        return ArgType.FILTER
+HINT = "Use '--' to separate paths from revisions, like this:\n'kart <command> [<revision>...] -- [<filter>...]'"
+SILENCING_HINT = "To silence this warning, use '--' to separate paths from revisions, like this:\n'kart <command> [<revision>...] -- [<filter>...]'\n"
 
-    raise click.UsageError(
-        f"Argument not recognised as a valid commit, ref, or range: {arg}"
-    )
 
+def _is_revision(repo, arg, warning_set):
+    # These things *could* be a path, but in that case the user should add a `--` before this arg to
+    # disambiguate, and they haven't done that here.
+    if arg == "[EMPTY]" or ".." in arg or HEAD_PATTERN.match(arg) or arg.endswith("^?"):
+        return True
 
-def parse_commits_and_filters(
+    filter_path = arg.split(":", maxsplit=1)[0]
+    head_tree = repo.head_tree
+    is_useful_filter_at_head = head_tree and filter_path in head_tree
```

Comment:
how's `filter_path in head_tree` work? It looks like `filter_path` is something like `["<datasetname>", "<pk_value>"]` and you're passing it to a `pygit2.Tree.__contains__`, which doesn't know about our path encodings. Or am I missing something?

Reply:
I only pass the first part of a filter to it - everything before the ":" - which is a path with no particular encoding. Everything after that is ignored. So if you have an arg like "points:feature:25" and a dataset called "points" and no revision called "points", we're happy to assume your arg is a filter - even without encoding the feature PK to see if it exists, it's enough that the dataset exists.

Reply:
good point, I missed L102